### PR TITLE
shield rework hotfix

### PR DESCRIPTION
**Pull request by SethDGamre**
solves nil error spam when playing scavengers

Fixes nil error on line 308 due to some units not having a default damage type

fixes exemptions in alldefs_post not actually being exemptions to shield damage protection

fixed chunk processing to actually process in chunks

### CODE REVIEW

**Review by WatchTheFort — APPROVED** [review]

**Comment by SethDGamre** [comment]
one more error to squash before night's out

File: luarules/gadgets/unit_shield_behaviour.lua [hunk 1]
```diff
@@ -304,8 +302,8 @@ function gadget:ShieldPreDamaged(proID, proOwnerID, shieldWeaponNum, shieldUnitI
 	-- proID isn't nil if hitscan weapons are used, it's actually -1.
 	if proID > -1 then
 		weaponDefID = projectileDefIDCache[proID] or spGetProjectileDefID(proID)
-
-		shieldData.shieldDamage = shieldData.shieldDamage + originalShieldDamages[weaponDefID]
+		local newShieldDamage = originalShieldDamages[weaponDefID] or 0 -- 0 to prevent rare bug where originalShieldDamages[weaponDefID] returns nil
```
Comment (sprunk) [inline luarules/gadgets/unit_shield_behaviour.lua:305]: why not just fill the missing part of the table?

Comment (WatchTheFort) [reply luarules/gadgets/unit_shield_behaviour.lua:305]: Wait a minute, is this from weapons whose damages table contains neither a `shields` entry nor a `default` entry?
https://github.com/beyond-all-reason/Beyond-All-Reason/blob/1284e003b762f6fefa339dff4eb4bb85597018fd/gamedata/alldefs_post.lua#L1396
e.g. AA weapons doing no damage to shields

Comment (sprunk) [reply luarules/gadgets/unit_shield_behaviour.lua:305]: the damage table is supposed to contain entries for all armor classes, even if left unspecified in the def file. the `or default` part should never even be reached

Comment (SethDGamre) [reply luarules/gadgets/unit_shield_behaviour.lua]: alldefs happens before those tables are populated, so it was coming up as nil for vtol damage types that don't have a default defined
I fixed this error.

File: gamedata/alldefs_post.lua [hunk 2]
```diff
@@ -1454,31 +1454,30 @@ function WeaponDef_Post(name, wDef)
...
+					wDef.customparams.shield_damage = wDef.damage.default
+				elseif wDef.damage.vtol then
+					wDef.customparams.shield_damage = wDef.damage.vtol * vtolShieldDamageMultiplier
+				else
+					wDef.customparams.shield_damage = 0
 				end
-			end
-
-			if wDef.damage ~= nil and wDef.damage.default ~= nil then
 				wDef.customparams = wDef.customparams or {}
```
Comment (WatchTheFort) [inline gamedata/alldefs_post.lua:1479]: This line is redundant, it already happened on line 1469

File: luarules/gadgets/unit_shield_behaviour.lua [hunk 3]
```diff
@@ -244,36 +247,35 @@ function gadget:GameFrame(frame)
...
-		local shieldUnitID = shieldUnitIndex[i]
-		local shieldData = shieldUnitsData[shieldUnitID]
-
-		if shieldData then
-			if not shieldData.shieldCoverageChecked then
-				if shieldData.shieldEnabled then
-					setCoveredUnits(shieldUnitID)
-				else
-					removeCoveredUnits(shieldUnitID)
+	if frame % 11 == 7 then
```
Comment (WatchTheFort) [inline luarules/gadgets/unit_shield_behaviour.lua:252]: What is the reason for offsetting by 7 frames? Is there other processing that runs every 11 frames that you don't want to sync up with?

File: gamedata/alldefs_post.lua [hunk 4]
```diff
@@ -1454,31 +1454,30 @@ function WeaponDef_Post(name, wDef)
...
-					break
+			if wDef.damage ~= nil then
+				-- Due to the engine not handling overkill damage, we have to store the original shield damage values as a customParam for unit_shield_behavior.lua to reference
+				wDef.customparams = wDef.customparams or {}
+				if wDef.damage.shields then
+					wDef.customparams.shield_damage = wDef.damage.shields
+				elseif wDef.damage.default then
+					wDef.customparams.shield_damage = wDef.damage.default
+				elseif wDef.damage.vtol then
+					wDef.customparams.shield_damage = wDef.damage.vtol * vtolShieldDamageMultiplier
```
Comment (WatchTheFort) [inline gamedata/alldefs_post.lua:1475]: This is not a good way to deal with weapons having no defined `shields` or `default` damage entries. This is fine as a stopgap measure, but the balance team will have to add `shields` damages entries for all the weapons that need them.

Comment (SethDGamre) [reply gamedata/alldefs_post.lua:1475]: Stopgap it is. Agreed.

It also will catch people not defining a shield or default damage which implies to me they don't want it to damage shields.

File: gamedata/alldefs_post.lua [hunk 5]
```diff
@@ -1454,31 +1454,30 @@ function WeaponDef_Post(name, wDef)
...
-			}
-
 			-- For balance, paralyzers need to do reduced damage to shields, as their raw raw damage is outsized
 			local paralyzerShieldDamageMultiplier = 0.25
 
-			if wDef.shield then
-				wDef.shield.repulser = false
-				wDef.shield.exterior = true
-			end
+			-- VTOL's may or may not do full damage to shields if not defined in weapondefs
```
[the remark this answers is not in the document]
Comment (SethDGamre) [reply gamedata/alldefs_post.lua:1460]: comment is accurate
"elseif wDef.damage.vtol then
wDef.customparams.shield_damage = wDef.damage.vtol * vtolShieldDamageMultiplier"
else
wDef.customparams.shield_damage = 0
end

**Pull request merged**